src/main/java/org/zalando/planb/provider/AuthorizeController.java

```diff
@@ -76,97 +130,273 @@ String showAuthorizationForm(@RequestParam(value = "realm") Optional<String> rea
+                                                          "invalid_request", "Missing redirect_uri")));
 
         validateRedirectUri(realmName, clientId, clientData, redirectUri);
 
-        model.addAttribute("responseType", responseType);
-        model.addAttribute("realm", realmName);
-        model.addAttribute("clientId", clientId);
-        model.addAttribute("scope", scope.orElse(""));
-        model.addAttribute("state", state.orElse(""));
-        model.addAttribute("redirectUri", redirectUri.toString());
-        model.addAttribute("error", error.orElse(null));
-
-        return "login";
-    }
-
-    @RequestMapping(value = "/oauth2/authorize", method = RequestMethod.POST)
-    void authorize(
-            @RequestParam(value = "response_type", required = true) String responseType,
-            @RequestParam(value = "realm") Optional<String> realmNameParam,
-            @RequestParam(value = "client_id", required = true) String clientId,
-            @RequestParam(value = "scope") Optional<String> scope,
-            @RequestParam(value = "redirect_uri") URI redirectUri,
-            @RequestParam(value = "state") Optional<String> state,
-            @RequestParam(value = "username", required = true) String username,
-            @RequestParam(value = "password", required = true) String password,
-            @RequestHeader(value = "Host") Optional<String> hostHeader,
-            HttpServletResponse response) throws IOException, URISyntaxException, JOSEException {
+        updateModelForLogin(model, responseType, realmName, clientId, scope, state, redirectUri, error);
+
+        return LOGIN_FORM;
+    }
+
+    @RequestMapping(method = RequestMethod.POST, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    AuthorizeResponse authorizeAsJson(
+            @RequestParam(value = RESPONSE_TYPE_SNAKECASE, required = true) final String responseType,
+            @RequestParam(value = REALM) final Optional<String> realmNameParam,
+            @RequestParam(value = CLIENT_ID_SNAKECASE, required = true) final String clientId,
+            @RequestParam(value = SCOPE) final Optional<String> scope,
+            @RequestParam(value = REDIRECT_URI_SNAKECASE) final URI redirectUri,
+            @RequestParam(value = STATE) final Optional<String> state,
+            @RequestParam(value = USERNAME, required = true) final String username,
+            @RequestParam(value = PASSWORD, required = true) final String password,
+            @RequestParam(value = DECISION) final Optional<String> decision,
+            @RequestHeader(value = HOST) final Optional<String> hostHeader) throws IOException, URISyntaxException,
+        JOSEException {
 
         if (!SUPPORTED_RESPONSE_TYPES.contains(responseType)) {
-            throw new BadRequestException("Unsupported response_type", "unsupported_response_type", "Unsupported response_type");
+            throw new BadRequestException("Unsupported response_type", "unsupported_response_type",
```

Comment:
this, along side with the exception being thrown in showAuthorizationForm feels like a duplicate.

I think we can solve it via polymorphism